### PR TITLE
Git web hook task

### PR DESCRIPTION
**Pull request by sanyokerem**

### CODE REVIEW

File: app/src/Command/CheckGitRepositoryUpdatesCommand.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Process\Process;
+
+class CheckGitRepositoryUpdatesCommand extends Command
+{
+    protected function configure()
+    {
+        $this
+            ->setName('app:git:check-updates')
```
Comment (makasim) [inline app/src/Command/CheckGitRepositoryUpdatesCommand.php]: app:split-repo

File: app/src/Controller/WebHookController.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+use Symfony\Component\Routing\Annotation\Route;
+
+class WebHookController extends Controller
+{
+    /**
+     * @Route("/test-webhook/{token}", name="test_web_hook", defaults={"token":""}, methods={"POST"})
+     */
+    public function testWebHook($token)
+    {
+        if ($token == '9WPTDSmLAWcmwcTPnFgKSMjl3qhigqLMKo9KB1ne') {
```
Comment (makasim) [inline app/src/Controller/WebHookController.php]: move it to parameters.yml

File: app/src/Controller/WebHookController.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+            $fs = new Filesystem();
+            $fs->touch('/var/git-web-hook');
+        }
+
+        return $this->json([]);
```
Comment (makasim) [inline app/src/Controller/WebHookController.php]: return a 204 response

File: app/src/Controller/WebHookController.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+namespace App\Controller;
+
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Routing\Annotation\Route;
+
+class WebHookController extends Controller
+{
+    /**
+     * @Route("/test-webhook/{token}", name="test_web_hook", defaults={"token":""}, methods={"POST"})
```
Comment (makasim) [inline app/src/Controller/WebHookController.php]: /hook/split-repo

**Comment by makasim** [comment]
Is there a way to find out what branch the code was updated on? From the (webhook)

File: app/src/Controller/WebHookController.php [hunk 5]
```diff
@@ -5,20 +5,21 @@
...
 class WebHookController extends Controller
 {
     /**
-     * @Route("/test-webhook/{token}", name="test_web_hook", defaults={"token":""}, methods={"POST"})
+     * @Route("/hook/split-repo/{token}", name="test_web_hook", defaults={"token":""}, methods={"POST"})
      */
     public function testWebHook($token)
     {
-        if ($token == '9WPTDSmLAWcmwcTPnFgKSMjl3qhigqLMKo9KB1ne') {
+        if ($token == $this->getParameter('git_web_hook_token')) {
```
Comment (makasim) [inline app/src/Controller/WebHookController.php]: ===

File: app/src/Controller/WebHookController.php [hunk 6]
```diff
@@ -5,20 +5,21 @@
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 use Symfony\Component\Filesystem\Filesystem;
 use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Flex\Response;
 
 class WebHookController extends Controller
 {
     /**
-     * @Route("/test-webhook/{token}", name="test_web_hook", defaults={"token":""}, methods={"POST"})
+     * @Route("/hook/split-repo/{token}", name="test_web_hook", defaults={"token":""}, methods={"POST"})
```
Comment (makasim) [inline app/src/Controller/WebHookController.php]: there is no default value for token

**Comment by sanyokerem** [comment]
> Is there a way to find out what branch the code was updated on? From the (webhook)
Yes. We can find out the branch from "ref" field in the request data.

File: apps/web-hook/src/Command/CheckGitRepositoryUpdatesCommand.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace App\Command;
+
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Process\Process;
+
+class CheckGitRepositoryUpdatesCommand extends Command
```
Comment (makasim) [inline apps/web-hook/src/Command/CheckGitRepositoryUpdatesCommand.php]: SplitRepoCommand

File: apps/web-hook/src/Controller/WebHookController.php [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    /**
+     * @Route("/hook/split-repo/{token}", name="test_web_hook", methods={"POST"})
+     */
+    public function testWebHook($token)
+    {
+        if ($token === $this->getParameter('git_web_hook_token')) {
+
+            $fs = new Filesystem();
+            $fs->touch('/var/git-web-hook');
+        }
```
Comment (makasim) [inline apps/web-hook/src/Controller/WebHookController.php]: if token do not match return 401

File: docker-compose.yml [hunk 9]
```diff
@@ -1,41 +1,47 @@
-version: "2"
+version: '2'
```
Comment (makasim) [inline docker-compose.yml]: 3

File: docker-compose.yml [hunk 10]
```diff
@@ -1,41 +1,47 @@
...
-      context: .
-      dockerfile: Dockerfile-nginx
-    volumes:
-     - ./app:/var/www/project/
-    ports:
-     - 80:80 # Ports that are exposed, you can connect to port 8001 to port 80 of the container.
-    networks:
-     - symfony
+    - './apps/web-hook:/jobs_dev/apps/web-hook:cached'
+    - './pkg/php-shared:/jobs_dev/apps/web-hook/vendor/comrade/shared:cached'
```
Comment (makasim) [inline docker-compose.yml]: there is no such folder, please remove

File: docker-compose.yml [hunk 11]
```diff
@@ -1,41 +1,47 @@
...
-    build:
-      context: .
-      dockerfile: Dockerfile-nginx
-    volumes:
-     - ./app:/var/www/project/
-    ports:
-     - 80:80 # Ports that are exposed, you can connect to port 8001 to port 80 of the container.
-    networks:
-     - symfony
+    - './apps/web-hook:/jobs_dev/apps/web-hook:cached'
```
Comment (makasim) [inline docker-compose.yml]: remove `:cached` cuz it does not work on linux.

File: docker-compose.yml [hunk 12]
```diff
@@ -1,41 +1,47 @@
...
 
-volumes:
-  symfony_db:
+  mongo:
+    image: 'mongo:3'
+    restart: 'always'
+    ports:
+    - "27017:27017"
+    volumes:
+    - './shared:/shared'
```
Comment (makasim) [inline docker-compose.yml]: not need in this project.

File: docker-compose.yml [hunk 13]
```diff
@@ -1,41 +1,47 @@
...
+    - VIRTUAL_HOST=web-hook.loc
+    - NGINX_WEB_ROOT=/jobs_dev/apps/web-hook/public
+    - NGINX_PHP_FALLBACK=/index.php
+    - NGINX_PHP_LOCATION=^/index\.php(/|$$)
+    - APP_DEV_PERMITTED=1
+    - SHELL_VERBOSITY=3
 
-  db:
-    image: mysql:5.6
+  rabbitmq:
```
Comment (makasim) [inline docker-compose.yml]: rabbitmq is not used

File: composer.json [hunk 14]
```diff
@@ -2,16 +2,34 @@
...
         "symfony/framework-bundle": "*",
+        "symfony/monolog-bundle": "^3.1",
+        "symfony/orm-pack": "*",
+        "symfony/process": "*",
+        "symfony/security-bundle": "*",
+        "symfony/serializer-pack": "*",
+        "symfony/swiftmailer-bundle": "^3.1",
+        "symfony/twig-bundle": "*",
+        "symfony/validator": "*",
+        "symfony/web-link": "*",
```
Comment (makasim) [inline composer.json]: make sure you installed only things you really need

File: config/bundles.php [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+    Doctrine\Bundle\DoctrineBundle\DoctrineBundle::class => ['all' => true],
+    Doctrine\Bundle\MigrationsBundle\DoctrineMigrationsBundle::class => ['all' => true],
+    Symfony\Bundle\SecurityBundle\SecurityBundle::class => ['all' => true],
+    Symfony\Bundle\SwiftmailerBundle\SwiftmailerBundle::class => ['all' => true],
+    Symfony\Bundle\WebProfilerBundle\WebProfilerBundle::class => ['dev' => true, 'test' => true],
+    Symfony\Bundle\TwigBundle\TwigBundle::class => ['all' => true],
+    Symfony\Bundle\MonologBundle\MonologBundle::class => ['all' => true],
+    Symfony\Bundle\DebugBundle\DebugBundle::class => ['dev' => true, 'test' => true],
+    Symfony\Bundle\MakerBundle\MakerBundle::class => ['dev' => true],
+    Symfony\Bundle\WebServerBundle\WebServerBundle::class => ['dev' => true],
```
Comment (makasim) [inline config/bundles.php]: make sure you registered bundle that you need, and remove all that is not needed.

File: docker-compose.yml [hunk 16]
```diff
@@ -1,41 +1,31 @@
...
-      - DATABASE_URL=mysql://root:root@db:3306/project_db # Connection string for the database.
-    volumes:
-     - ./app:/var/www/project/ # Location of the project for php-fpm. Note this should be the same for NGINX.
-    networks:
-     - symfony # Docker containers (services) that need to connect to each other should be on the same network.
-
-  nginx:
+  web-hook:
+    restart: 'always'
+    working_dir: '/jobs_dev'
```
Comment (makasim) [inline docker-compose.yml]: /app

File: Dockerfile [hunk 17]
```diff
@@ -0,0 +1,4 @@
+FROM formapro/nginx-php-fpm:latest-all-exts
+
+RUN chmod 757 /var
+RUN apt-get update && apt-get install -y git
```
Comment (makasim) [inline Dockerfile]: what do you need git for?

File: src/Command/SplitRepoCommand.php [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+            ->setDescription('Check if there has been some changes on a GitHub repository');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $fs = new Filesystem();
+
+        $output->writeln('Checking ...');
+
+        if (!$fs->exists('/var/git-web-hook')) {
```
Comment (makasim) [inline src/Command/SplitRepoCommand.php]: you should use app var folder not a system one

File: src/Command/SplitRepoCommand.php [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+            $output->writeln('There have been no updates');
+
+            return;
+        }
+
+        $branch = file_get_contents('/var/git-web-hook');
+
+        if ($fs->exists('/var/repository')) $fs->remove('/var/repository');
```
Comment (makasim) [inline src/Command/SplitRepoCommand.php]: it is not psr1\2 compliant code. Please learn them and use from now on

File: src/Command/SplitRepoCommand.php [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+
+        $process->setWorkingDirectory('/var/repository');
+        $process->setCommandLine(sprintf('git branch %1$s; git checkout %1$s', $branch));
+        $process->run();
+        $process->setCommandLine('ls -l');
+        $process->run();
+        $output->writeln($process->getOutput());
+        $process->setCommandLine($commandFile);
+        $process->start();
+        $process->wait(function ($type, $buffer) use ($output) {
```
Comment (makasim) [inline src/Command/SplitRepoCommand.php:88]: is this some kind of debug code?

File: Dockerfile (hunk 17 above)
Comment (sanyokerem) [reply Dockerfile]: For using git clone and push commands

File: src/Command/SplitRepoCommand.php (hunk 20 above)
Comment (sanyokerem) [reply src/Command/SplitRepoCommand.php:88]: Yes

File: .env.dist [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+# Format described at http://docs.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/configuration.html#connecting-using-a-url
+# For an SQLite database, use: "sqlite:///%kernel.project_dir%/var/data.db"
+# Configure your db driver and server_version in config/packages/doctrine.yaml
+DATABASE_URL=mysql://db_user:db_password@127.0.0.1:3306/db_name
+
+
+# For Gmail as a transport, use: "gmail://username:password@localhost"
+# For a generic SMTP server, use: "smtp://localhost:25?encryption=&auth_mode="
+# Delivery is disabled by default via "null://localhost"
+MAILER_URL=null://localhost
```
Comment (makasim) [inline .env.dist:20]: remove it

File: Dockerfile [hunk 22]
```diff
@@ -0,0 +1,3 @@
+FROM formapro/nginx-php-fpm:latest-all-exts
+
+RUN apt-get update && apt-get install -y git
```
Comment (makasim) [inline Dockerfile:3]: do apt update as here https://github.com/formapro/docker-nginx-php-fpm/blob/master/base/Dockerfile#L7

File: composer.json [hunk 23]
```diff
@@ -2,12 +2,13 @@
     "type": "project",
     "license": "proprietary",
     "require": {
-        "php": "^7.0.8",
+        "php": "^7.1.3",
         "ext-ctype": "*",
         "ext-iconv": "*",
         "symfony/console": "*",
-        "symfony/flex": "^1.1",
```
Comment (makasim) [inline composer.json:9]: flex is needed

File: composer.json [hunk 24]
```diff
@@ -2,12 +2,13 @@
     "type": "project",
     "license": "proprietary",
     "require": {
-        "php": "^7.0.8",
+        "php": "^7.1.3",
         "ext-ctype": "*",
         "ext-iconv": "*",
         "symfony/console": "*",
```
Comment (makasim) [inline composer.json:8]: `^4`

File: config/services.yaml [hunk 25]
```diff
@@ -4,6 +4,10 @@
 # Put parameters here that don't need to change on each machine where the app is deployed
 # https://symfony.com/doc/current/best_practices/configuration.html#application-related-configuration
 parameters:
+    app.git_web_hook_token: 9WPTDSmLAWcmwcTPnFgKSMjl3qhigqLMKo9KB
+    app.git_repo_link: git@github.com:sanyokerem/enqueue-dev.git
+    app.command_file_path: /bin/subtree-split
+    app.hook_dir: '%kernel.project_dir%/var/split_repo'
```
Comment (makasim) [inline config/services.yaml:10]: The hook file should contain a branch name so different branches don't interfere with each other.

File: docker-compose.yml [hunk 26]
```diff
@@ -1,41 +1,21 @@
...
-
-  db:
-    image: mysql:5.6
+    - './:/app'
+    env_file: '.env'
     environment:
-    - MYSQL_ROOT_PASSWORD=root # Setting the MYSQL credentials to root:root.
-    volumes:
-    - symfony_db:/var/lib/mysql:cached # Persist the database in a Docker volume.
+    - VIRTUAL_HOST=web-hook.loc
```
Comment (makasim) [inline docker-compose.yml:14]: remove, not used

File: src/Command/SplitRepoCommand.php [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+        $branch = file_get_contents($this->container->getParameter('app.hook_dir') . '/git-web-hook');
+
+        if ($fs->exists($this->container->getParameter('app.hook_dir') . '/repository')) {
+            $fs->remove($this->container->getParameter('app.hook_dir') . '/repository');
+        }
+
+        $output->writeln('Repository: ' . $this->container->getParameter('app.git_repo_link'));
+
+        $process = new Process(sprintf(
+            'git clone %1$s -b %2$s %3$s/repository',
```
Comment (makasim) [inline src/Command/SplitRepoCommand.php:56]: git clone %s -b %s %s/repository

File: src/Command/SplitRepoCommand.php [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        ));
+        $process->start();
+        $process->wait(function ($type, $buffer) use ($output) {
+            if (Process::ERR === $type) {
+                $output->write($buffer);
+            } else {
+                $output->write($buffer);
+            }
+        });
+
```
Comment (makasim) [inline src/Command/SplitRepoCommand.php:69]: you have to check exit status at this stage.

File: src/Command/SplitRepoCommand.php [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+        $fs->remove($this->container->getParameter('app.hook_dir') . '/repository');
+        $fs->remove($this->container->getParameter('app.hook_dir') . '/git-web-hook');
```
Comment (makasim) [inline src/Command/SplitRepoCommand.php:97]: do it at the beginning .

File: src/Controller/WebHookController.php [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        if (!$token === $this->getParameter('app.git_web_hook_token')) {
+            return new Response('', 401);
+        }
+
+        if ($request->headers->get('X-GitHub-Event') === 'ping') {
+            return new Response('');
+        }
+
+        $requestContent = json_decode($request->getContent());
```
Comment (makasim) [inline src/Controller/WebHookController.php:22]: make sure it is an object, if not return 400.

File: src/Controller/WebHookController.php [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+
+        if (!isset($requestContent->ref)) {
+            return new Response('', 404);
```
Comment (makasim) [inline src/Controller/WebHookController.php:25]: 400

**Pull request merged**